go-controller/vendor/github.com/coreos/go-iptables/iptables/iptables.go

```diff
@@ -185,6 +220,43 @@ func (ipt *IPTables) Insert(table, chain string, pos int, rulespec ...string) er
 	return ipt.run(cmd...)
 }
 
+// Replace replaces rulespec to specified table/chain (in specified pos)
+func (ipt *IPTables) Replace(table, chain string, pos int, rulespec ...string) error {
+	cmd := append([]string{"-t", table, "-R", chain, strconv.Itoa(pos)}, rulespec...)
+	return ipt.run(cmd...)
+}
+
+// InsertUnique acts like Insert except that it won't insert a duplicate (no matter the position in the chain)
+func (ipt *IPTables) InsertUnique(table, chain string, pos int, rulespec ...string) error {
+	exists, err := ipt.Exists(table, chain, rulespec...)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		return ipt.Insert(table, chain, pos, rulespec...)
+	}
+
+	return nil
+}
+
+// Restore replaces specified chains and rules in a specific table
+// rulesMap is keyed by chain name, and holds slices of rulespecs
+// Only chains specified in the map will be flushed and replaced. Other chains will not be affected.
+func (ipt *IPTables) Restore(table string, rulesMap map[string][][]string) error {
+	restoreRules := "*" + table
+	for chain, rules := range rulesMap {
```

Comment:
I don't see you checking if the chains exist or not here and also its not checked when we call this from ovnkube, is that intended?

Reply:
right, theres no reason to because they will all just be recreated on restore. Its an atomic replacement of any chain listed in the restore input.

Reply:
ack IIUC you are also doing the:
```
:OVN-KUBE-ETP - [0:0]
```
create chain as part of restore right?
the only thing you don't do is the jump to that chain main iptable chains while restoring? and that is a separate thing?
I wonder if when you restore let's say OVN-KUBE-ETP chain if the `[1246:214409] -A PREROUTING -j OVN-KUBE-ETP` => can just exist without issues even if that chain is temporarily recreated totally?

Reply:
good question. I tested it and it does not remove the other rules referencing it. In fact I cant even delete the the chain itself until I remove the reference to it or I get the error: iptables: Too many links.

Reply:
ah ack I see that's good as long as we don't try to cleanup a referenced one.